lib/spells/AdventureSpellMechanics.cpp

```diff
@@ -275,15 +280,15 @@ ESpellCastResult DimensionDoorMechanics::applyAdventureEffects(SpellCastEnvironm
 		env->complain("Destination is out of map!");
 		return ESpellCastResult::ERROR;
 	}
-	
-	if(!parameters.caster->getHeroCaster())
+
+	if(parameters.caster->getAdventureCasterID() < 0)
```

Comment:
caster->isAdventureCaster() or even caster->canCastAdventureSpells()

Reply:
When I think more about it I can not see a way to cast dimension door with anything except a hero. So looks like we need a way to retrive CGHeroInstance from a caster. So this check for adventure caster < 0 sounds weird. If we need a hero we should ask if it is a hero. Otherwise we need an abstraction which can answer how many caster have mana, movement points, sight center and so on.

Reply:
Now it is not possible to make some other adventure entity able to cast dimension door because we expect hero here. And this abstraction with ID will not help

Reply:
Why shipyard needs to cast an adventure spell? Can't it just build a boat? Shipyard mechanics is a bit different from boat summoning. One needs resources, another needs mana. Also shipyard never summons existing boat no matter if it exists or not. Mixing up such things may be dangerous. So still they have something in common. Some boat placement mechanics.
Agree, this is a complex question if we should use hero instance or some abstract id.

Reply:
Anyway if you want a strong abstraction you should always talk to caster itself and never cast it to hero, object or whatever. If you do not want strong abstraction and a lot of proxy methods like getCasterMana, consumeCasterMana or whatever you may have a set of methods to get casterAsBattleUnit, casterAsHero or casterAsObject and so on. some kind of dynamic cast. You cast to required instance and work with it. IF it is null and required you report an error.

Reply:
A very complex but interesting solution, Spell creates a spell mechanic factory and gives it to caster, caster calls createMechanic on this factory and pass this in it. Depending on who caster is factory has a few overloads for hero, object, battle unit. Factory creates mechanic which has all exact objects inside. IF mechanics needs to be able to  work with both hero and adv object it can have two subclasses with virtual hooks. For instance SummonBoatMechanicsFactory can return SummonBoatMechanics if called for hero and BuildBoatMechanics if called for shipyard.
So factory interface should know about all possible casters and in case if some caster is not supported it can crash or return nullptr. In last case methods should be named like tryCreateMechanics.
This will allow mechanics to work with exact types of casters or build local abstractions on them.
So this means local abstractions may get duplicated.

Reply:
How about composite mechanics? We can decompose mechanics onto smaller parts like ManaConsumptionMechanics, MovementConsumptionMechanics, and so on. This will allow combining mechanics and even constructing composite spells like Meteor Shower with wall destruction. Depending on caster some mechanics can get replaced with others. This will remove duplication of mechanics code for factory solution

Reply:
Ok, these were just some ideas, they are to complex to implement.

Reply:
Your controversal commit is moved outside of this PR.